js/utils/data_utils/XLSX_handlers.js

```diff
@@ -53,13 +31,14 @@ function deleteTopRowsUntilFullData(data) {
     return data;
 }
 
-function processWorkbook(workbook) {
+export function processWorkbook(arrayBuffer) {
+    const workbook = XLSX.read(arrayBuffer, { type: 'array' });
     workbook.SheetNames.forEach(sheetName => {
         // only convert sheets we need
         if (Object.keys(SHEETS).includes(sheetName)) {
              // read in sheets
             const sheet = workbook.Sheets[sheetName];
-            const rawData = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: null });
+            const rawData = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: '' });
```

Comment:
> the initial Excel file read-in is a bit slow 

It's hard to say if it's the arraybuffer read / Blob write or JSON serialization / deserialization that is slowing things down without profiling the code. How bad is the performance? How does it scale with the size of the file / data? 

> Local storage doesn't seem perfect for this, ... , but I think it might be the best (only?) way to do it with a pure front-end solution

If the performance hit is in the JSON serialization / deserialization, then there is the option to use [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB) which is a client-side NoSQL style storage. Then we could store the data as a binary blob instead of JSON. That would have a performance improvement but require some extra effort to setup.

Is the performance hit worth addressing in your opinion?

Reply:
It's not that slow -- a couple seconds. Based on a quick look into IndexedDB, it's probably not worth changing the approach for now, but I'll test the current code to see how long it takes to process some of the larger departments.

Reply:
Yea, I agree. A couple seconds isn't bad. A better, lower-effort fix would probably be to handle the IO and serialization operations as asynchronous operations and provide some UX treatment like a spinner while they process.